### PR TITLE
Close connections on worker on shutdown, add force shutdown

### PR DESCRIPTION
**Pull request by vancluever**
This adds functionality to ensure that connections are closed on a
worker when the server is shut down or reloaded.

Connections are closed, and then the shutdown process waits for the next
report to the controller to go through, waiting the grace period that
was added in #1330. The rest of shutdown is then proceeded with.
Listener shutdown has been moved to the start of the process to ensure
that new connections can't be made while we are shutting down.

Additionally, functionality has been added to force shut down servers
after a second interrupt has been received after the first one. This
takes the form of code in both the server and dev mode code paths that
just performs an os.Exit if another shutdown is received during the
normal shutdown process.

Finally, the commit also includes some refactoring to tests so that we
could use some of the new E2E testing code to test session connections
during shutdown and reload.

### CODE REVIEW

**Comment by vancluever** [comment]
Screenshots for the forced shutdown:

## `boundary server`

image

## `boundary dev`

(With debug logging to demonstrate gap between initial CTRL-C and forced shutdown)

image

**Comment by malnick** [comment]
Linking related issue: https://github.com/hashicorp/boundary/issues/894

**Comment by vancluever** [comment]
rebased off of #1340

File: internal/cmd/commands/dev/dev.go
Comment (jefferai) [inline internal/cmd/commands/dev/dev.go:578]: I think this is missable if the second interrupt arrives prior to this being registered. I don't have a great idea for how to do it better, and it's unlikely to happen since the second would have to come _very_ quickly. Just calling it out in case you have other ideas.

Comment (vancluever) [reply internal/cmd/commands/dev/dev.go:578]: Yeah, none short of rewriting [`MakeShutdownCh`](https://github.com/hashicorp/boundary/blob/4ba7eb5ff44d9a1541adf42f70d85541ed82ce28/internal/cmd/base/base.go#L106-L119). I'll noodle on it for a few days.

**Comment by vancluever** [comment]
Going to hold off on fixing conflicts for now until #1400 is merged. After that there will probably be a reabase and squash here so that we can track code that will be moving to the `helper` package.

**Comment by vancluever** [comment]
As mentioned, this has been rebased and squashed and should be ready for review now!

File: internal/tests/helper/testing_helper.go
Comment (jefferai) [inline internal/tests/helper/testing_helper.go:28]: Can you repoint this to `github.com/hashicorp/go-secure-stdlib/base62` and `go mod tidy` after?

File: internal/cmd/commands/dev/dev.go
Comment (jefferai) [reply internal/cmd/commands/dev/dev.go:573]: Resolving for now, I don't have a better idea and I think it's a very unlikely scenario.

File: internal/servers/worker/worker.go
Comment (jefferai) [inline internal/servers/worker/worker.go:185]: This function is only used in one place, can this be inlined above?

**Comment by jefferai** [comment]
Looking good! Just two minor comments.

**Comment by vancluever** [comment]
@jefferai updated!

**Review by jefferai — APPROVED** [review]
Looks good pending fixing merge conflicts and running `go mod tidy`! Nice work!

**Review by jefferai — APPROVED** [review]

**Pull request merged**
